tombstones.go

```diff
@@ -48,24 +48,41 @@ func setTombstone(shardPath string, repoID uint32, tombstone bool) error {
 	}
 
 	dest := shardPath + ".meta"
-	err = jsonMarshalMeta(repos, dest)
+	temp, err := JsonMarshalRepoMetaTemp(repos, filepath.Dir(dest), filepath.Base(dest)+".*.tmp")
 	if err != nil {
 		return err
 	}
 
+	err = os.Rename(temp, dest)
+	if err != nil {
+		os.Remove(temp)
+	}
+
 	return nil
 }
 
-func jsonMarshalMeta(v interface{}, p string) (err error) {
-	b, err := json.Marshal(v)
+// JsonMarshalRepoMetaTemp writes the JSON encoding of the repository metadata to a temporary file
+// in the directory dir, and returns the name of the created file.
+// The filename name is generated by taking pattern and adding a random string to the end.
+// If pattern includes a "*", the random string replaces the last "*".
+// If dir is the empty string, CreateTemp uses the default directory for temporary files, as returned by os.TempDir.
+// It is the caller's responsibility to remove the file when it is no longer needed.
+// TODO: Should we stick this in a util package?
+// TODO: Maybe we should give this the path of a shard, and it returns a 3 tuple of (finalFile, tempFile, error)?
+//       If we do this, we can have it automatically figure out whether it needs to write out []zoekt.Repository or just
+//       zoekt.Repository. This would result in duplicate calls to zoekt.ReadMetadataPath at some of the callsites though
+//       such as the one above.
+func JsonMarshalRepoMetaTemp(repositoryMetadata interface{}, dir, pattern string) (tempFile string, err error) {
```

Comment:
I'm not super happy with this function. We now have multiple places that need to write repository metadata (shard merging and the builder for delta shards), and there is some nuance here with the `chmod` logic that all all of these places need to maintain. I think a shared function is the answer to that, but something about the interface feels awkward to me.

Reply:
Perhaps combining this function with the version check information (https://github.com/sourcegraph/zoekt/pull/274/files#r813424230) would make this interface a bit deeper and more useful?

Reply:
I wanted to share this logic between zoekt-merge and the builder, which would require supporting both the normal and compund shard usecase. Maybe it's not worth it, however?

Reply:
I changed the signature of the function [here](https://github.com/sourcegraph/zoekt/pull/274/commits/8d3e8f853c7d02add4724ddff32a2d3adba238ac): https://github.com/sourcegraph/zoekt/blob/8d3e8f853c7d02add4724ddff32a2d3adba238ac/tombstones.go#L63-L70

If you still think it's not worth support both the simple and compound shard usecase with this function (the reason why the signature includes `interface{}`), I can just duplicate the logic and create a separate function for just the builder's use-case.

Reply:
I am ok with leaving it like this.